File: 15/p15.py
"""
project euler problem 15
"""

def path_counter(grid_size):
    """
    counts number of paths by starting at bottom of grid and moving up, adding
    possible paths until you reach 0,0 where you return
    """

    grid = [[0] * (grid_size+1)] * (grid_size+1)

    for i in range(grid_size):
        grid[i][grid_size] = 1
        grid[grid_size][i] = 1

    for row in range(grid_size-1, -1, -1):
        for col in range(grid_size-1, -1, -1):
            grid[row][col] = grid[row+1][col] + grid[row][col+1]

    return grid[0][0]

def main():
    print(path_counter(20))

if __name__ == '__main__':
    main()

File: 15/p15.cpp
/*
project euler problem 15
*/
#include <iostream>

long pathCounter(const int gridSize){
  /*
  counts number of paths by starting at bottom of grid and moving up, adding
  possible paths until you reach 0,0 where you return
  */

  long grid[gridSize+1][gridSize+1];

  for (int i = 0; i < gridSize; i++) {
    grid[i][gridSize] = 1;
    grid[gridSize][i] = 1;
  }

  for (int row = gridSize-1; row >= 0; row--) {
    for (int col = gridSize-1; col >= 0; col--) {
      grid[row][col] = grid[row+1][col] + grid[row][col+1];
    }
  }

  return grid[0][0];

}

int main(int argc, char const *argv[]) {
  std::cout << pathCounter(20) << '\n';
  return 0;
}
